students/yushus/session06/test_mailroom4.py

```diff
@@ -0,0 +1,59 @@
+#!/usr/bin/env/python3
+
+"""
+Yushu Song
+Au2018-Py210B
+Mailroom4 assignment
+"""
+
+import os
+import pytest
+import re
+import sys
+import shutil
+from mailroom4 import get_report, get_file_name, update_donor, update_donation, get_file_name, get_a_thank_you_letter
+
+def test_get_report_success():
+    donor_db = {"Bill Gates": [100, 200],"Jack Ma": [500, 500]}
+    assert get_report(donor_db) == [["Jack Ma", 1000, 2, 500],
+                                    ["Bill Gates", 300, 2, 150]]
+
+def test_update_donor_success():
+    donor_db = {"Bill Gates": [100, 200],"Jack Ma": [500, 500]}
+    update_donor("Yushu Song", donor_db)
+    assert "Yushu Song" in donor_db.keys()
+
+def test_update_donor_failure():
+    with pytest.raises(ValueError, match=r'Donor name.*'):
+        update_donor("", {})
+
+def test_update_donation_success():
+    donor_db = {"Bill Gates": [100, 200],"Jack Ma": [500, 500]}
+    update_donation("Bill Gates", 300, donor_db)
+    assert donor_db["Bill Gates"][-1] == 300
+
+def test_update_donation_failure():
+    donor_db = {"Bill Gates": [100, 200],"Jack Ma": [500, 500]}
+    with pytest.raises(ValueError, match=r'.*\s+0'):
+        update_donation("Bill Gates", 0, donor_db)
+
+def test_get_file_name_replace_char_success():
+    path = os.getcwd()
+    file_name = "Yushu*Song"
+    assert os.path.join(path, "Yushu_Song.txt") == get_file_name(path, file_name)
+
+def test_get_file_name_new_dir_success():
+    try:
+        path = os.path.join(os.getcwd(), 'temp')
+        file_name = "Yushu*Song"
+        assert os.path.join(path, "Yushu_Song.txt") == get_file_name(path, file_name)
+    except Exception as ex:
+        print(ex)
+    finally:
+        # Clean up resource afterwards
+        shutil.rmtree(path)
```

Comment:
Great to see that you are cleaning up! However, I'd say it's a little unexpected that the function `get_file_name()` creates a directory if it's not there yet. It seems kind of an unexpected side effect.

It'd have been better if the directory was created in the actual file creation function. I know that you didn't want to pass the directory path to that function, but just pass the full path of the file only, which requires the directory to be created, so I suppose that's why you created the directory somewhere, but I'd say `get_file_name()` is not a good place to create the directory. Also, you could have run the code that generates all the letter files and validates that those files are really generated, and after that, remove the entire directory. Just a thought. No need to change your code at this point.